src/security_utilities_rust/src/microsoft_security_utilities_core/identifiable_scans.rs

```diff
@@ -110,11 +138,136 @@ impl PossibleScanMatch {
             ScanMatchType::His32Utf16 => { HIS_32_UTF16_LEN },
             ScanMatchType::His64Utf8 => { HIS_64_UTF8_LEN },
             ScanMatchType::His64Utf16 => { HIS_64_UTF16_LEN },
+            ScanMatchType::His39Utf8 => { HIS_39_UTF8_LEN },
+            ScanMatchType::His39Utf16 => { HIS_39_UTF16_LEN },
+            ScanMatchType::His40Utf8 => { HIS_40_UTF8_LEN },
+            ScanMatchType::His40Utf16 => { HIS_40_UTF16_LEN },
+            ScanMatchType::HisA7Utf8 => { HIS_A7_UTF8_LEN },
+            ScanMatchType::HisA7Utf16 => { HIS_A7_UTF16_LEN },
+            ScanMatchType::HisA8Utf8 => { HIS_A8_UTF8_LEN },
+            ScanMatchType::HisA8Utf16 => { HIS_A8_UTF16_LEN },
         }
     }
 
     pub fn match_type(&self) -> &ScanMatchType { &self.mtype }
 
+    fn his_a7_matched_bytes(data: &[u8]) -> usize {
+        /*
+         * 3 Base64 + 3 signature + 31 Base64 + optional 1 [=]
+         */
+        if data.len() < HIS_A7_UTF8_LEN {
+            return 0;
+        }
+
+        for b in &data[0..3] {
+            if !b.is_url_unreserved() {
+                return 0;
+            }
+        }
+
+        for b in &data[6..37] {
+            if !b.is_url_unreserved() {
+                return 0;
+            }
+        }
+
+        HIS_A7_UTF8_LEN
+    }
+
+    fn his_a8_matched_bytes(data: &[u8]) -> usize {
+        /*
+         * 3 Base64 + 3 signature + 34 Base64 + optional 1 [=]
+         */
+        if data.len() < HIS_A8_UTF8_LEN {
+            return 0;
+        }
+
+        for b in &data[0..3] {
+            if !b.is_url_unreserved() {
+                return 0;
+            }
+        }
+
+        for b in &data[6..40] {
+            if !b.is_url_unreserved() {
+                return 0;
+            }
+        }
+
+        HIS_A8_UTF8_LEN
+    }
+
+    fn his_39_matched_bytes(data: &[u8]) -> usize {
+        /*
+         * 42 Base64 + 4 signature + 1 [A-D] + 5 Base64 + optional 1 [=]
+         */
+        if data.len() < 52 {
+            return 0;
+        }
+
+        for b in &data[0..42] {
+            if !b.is_base64() {
+                return 0;
+            }
+        }
+
+        /* NOTE: We skip the signature since we already checked */
+
+        if data[46] < b'A' || data[46] > b'D' {
+            return 0;
+        }
+
+        for b in &data[47..52] {
+            if !b.is_base64() {
+                return 0;
+            }
+        }
+
+        if data.len() >= HIS_39_UTF8_LEN {
+            if data[52] == b'=' {
+                return HIS_39_UTF8_LEN;
+            }
+        }
+
+        52
+    }
+
+    fn his_40_matched_bytes(data: &[u8]) -> usize {
```

Comment:
>his_40_matched_bytes [](http://example.com/codeflow?start=7&length=20)

Just a note, there are only 3 forms of matching for any HIS v1 encoded thing, which map to key forms require no, two or 4 bits of padding to keep the fixed signature lined up properly in the base64-encoded form. This helper captures one pattern, your 64 byte helper captures another and the 39 byte helper captures the third scenario. we may not require any generalization taking advantage of this fact, since the fact is we don't project any additional HIS v1 keys that aren't 32, 39, 40, or 64 byte in length.

Reply:
[a-zA-Z0-9\-_]{22}XXXX[A-D][a-zA-Z0-9\-_]{5}
[a-zA-Z0-9\-_]{24}XXXX[a-zA-Z0-9\-_]{5}[AQgw]
[a-zA-Z0-9\-_]{25}XXXX[A-P][a-zA-Z0-9\\-_]{5}